### PR TITLE
194 Added api to get blood donation posts

### LINKED ISSUE #194: Api to get blood donation post list
As a user, we need an API that retrieves the list of donation requests based on status. This endpoint will return information about donations, including details like the donor, blood group, donation status, location, and any filters related to donor availability.

### PR DESCRIPTION
**Pull request by mahinuzzaman**
Fixes #194 
- Added api to get blood donation posts
- Fixed Cognito schema miss match issue
- Update donation patch to get seekerId from authorizer

- [x] Does not depend on other PRs
- [ ] Depends on other PR being merged first: #

## What kind of change does this PR introduce (check at least one)

- [x] Bugfix
- [ ] Improvement
- [x] New feature
- [ ] Refactor
- [ ] Infrastructure
- [ ] Other (please specify)

## How would you describe the changes in a changelog (1-2 lines)?

Provide a brief overview of the changes introduced in this pull request.

## Related Issues/PRs (optional)

Link to the relevant issue(s) or provide a clear description of the problem or feature being addressed.

## How to test

- [ ] Needs regression testing (add QA needed label)

## Checklist

- [x] I have tested these changes locally and they function as expected.
- [ ] I have added or updated necessary documentation to reflect the changes.
- [ ] I have followed the project's coding style and guidelines.
- [ ] I have added appropriate test coverage to ensure the changes are functioning correctly.
- [x] I have checked for any potential conflicts with other branches or pull requests.
- [ ] I have updated necessary OpenAPI specs

## Reviewer Request (optional)

If you have any specific requests or suggestions for reviewers, mention them here.

## Screenshots (optional)

If applicable, provide any additional screenshots or GIFs that showcase the changes made in this pull request.

## Demo Link (optional)

If applicable, provide a link to a live demo or deployment where the changes can be reviewed.

## Known Issues (optional)

If there are any known issues or limitations with the changes made in this pull request, describe them here.

## Additional Notes (optional)

Add any additional notes or comments you think are relevant to this pull request.

## Deploy risk & follow up

Please assess risk & follow up on deployment of this PR. Elaborate your choice if selecting anything other than 'No risk'.

- [ ] No risk
- [x] Low
- [ ] High

### CODE REVIEW

**Review by sakibstark11 — DISMISSED** [review]

File: iac/terraform/aws/api_gw.tf [hunk 1]
```diff
@@ -34,22 +34,49 @@ resource "aws_api_gateway_deployment" "api_deployment" {
...
   lifecycle {
     create_before_destroy = true
   }
 }
 
+resource "aws_iam_policy" "api_gw_dynamodb_policy" {
+  name        = "${var.environment}-api-gw-dynamodb-policy"
+  description = "Policy allowing API Gateway to query DynamoDB tables"
+
+  policy = jsonencode({
```
Comment (melon-ruet) [inline iac/terraform/aws/api_gw.tf:50]: Do not use jsonencode. Use this
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/iam_policy_document

File: openapi/integration/aws/donations/get-donations.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
...lit('#')[1]\",\n \"reqPostId\": \"$item.SK.S.split('#')[1]\",\n \"location\": \"$item.location.S\",\n \"urgencyLevel\": \"$item.urgencyLevel.S\",\n \"status\": \"$item.status.S\",\n \"createdAt\": \"$item.createdAt.S\",\n \"contactInfo\": {\n \"name\": \"$item.contactInfo.M.name.S\",\n \"phone\": \"$item.contactInfo.M.phone.S\"\n },\n \"neededBloodGroup\": \"$item.neededBloodGroup.S\",\n \"bloodQuantity\": \"$item.bloodQuantity.N\",\n \"transportationInfo\": \"$item.transportationInfo.S\",\n \"donationDateTime\": \"$item.donationDateTime.S\",\n \"geohash\": \"$item.geohash.S\",\n \"longitude\": \"$item.longitude.N\",\n \"latitude\": \"$item.latitude.N\",\n \"acceptedDonors\": [\n #foreach($acceptedItem in $inputRoot.Items)\n #if($acceptedItem.SK.S.startsWith('ACCEPTED#') && $acceptedItem.PK.S == $item.PK.S)\n #set($acceptedReqPostID = $acceptedItem.SK.S.split('#')[1])\n #if($acceptedReqPostID == $item.SK.S.split('#')[1])\n {\n \"reqPostID\": \"$acceptedReqPostID\",\n \"donorULID\": \"$acceptedItem.SK.S.split('#')[2]\"\n }#if($foreach.hasNext),#end\n #end\n #end\n #end\n ]\n }#if($foreach.hasNext),#end\n #end\n #end\n  ],\n \"message\": \"Donations retrieved successfully\"\n}\n#end"
```
Comment (melon-ruet) [inline openapi/integration/aws/donations/get-donations.json:15]: This is not readable

File: openapi/integration/aws/donations/patch-donation.json [hunk 3]
```diff
@@ -16,7 +16,7 @@
...
+    "application/json": "#set($inputRoot = $input.path('$'))\r\n{\r\n  \"seekerId\": \"$context.authorizer.claims['custom:userId']\",\r\n  \"requestPostId\": \"$inputRoot.requestPostId\",\r\n  \"bloodQuantity\": $inputRoot.bloodQuantity\r\n  #if($inputRoot.urgencyLevel != '')\r\n  , \"urgencyLevel\": \"$inputRoot.urgencyLevel\"\r\n  #end\r\n  #if($inputRoot.donationDateTime != '')\r\n  , \"donationDateTime\": \"$inputRoot.donationDateTime\"\r\n  #end\r\n  #if($inputRoot.contactInfo.name != '' || $inputRoot.contactInfo.phone != '')\r\n  , \"contactInfo\": {\r\n    #set($comma = false)\r\n    #if($inputRoot.contactInfo.name != '')\r\n    \"name\": \"$inputRoot.contactInfo.name\"\r\n    #set($comma = true)\r\n    #end\r\n    #if($inputRoot.contactInfo.phone != '')\r\n    #if($comma), #end\r\n    \"phone\": \"$inputRoot.contactInfo.phone\"\r\n    #end\r\n  }\r\n  #end\r\n  #if($inputRoot.transportationInfo != '')\r\n  , \"transportationInfo\": \"$inputRoot.transportationInfo\"\r\n  #end\r\n  #if($inputRoot.shortDescription != '')\r\n  , \"shortDescription\": \"$inputRoot.shortDescription\"\r\n  #end\r\n}\r\n"
```
Comment (melon-ruet) [inline openapi/integration/aws/donations/patch-donation.json:19]: This is not readable

File: openapi/integration/aws/donations/post-donation.json [hunk 4]
```diff
@@ -16,7 +16,7 @@
...
+    "application/json": "#set($inputRoot = $input.path('$'))\n{\n  \"seekerId\": \"$context.authorizer.claims['custom:userId']\",\n  \"neededBloodGroup\": \"$inputRoot.neededBloodGroup\",\n  \"bloodQuantity\": $inputRoot.bloodQuantity,\n  \"urgencyLevel\": \"$inputRoot.urgencyLevel\",\n  \"location\": \"$inputRoot.location\",\n  \"latitude\": $inputRoot.latitude,\n  \"longitude\": $inputRoot.longitude,\n  \"donationDateTime\": \"$inputRoot.donationDateTime\",\n  \"contactInfo\": {\n    \"name\": \"$inputRoot.contactInfo.name\",\n    \"phone\": \"$inputRoot.contactInfo.phone\"\n  },\n  \"patientCondition\": \"$inputRoot.patientCondition\",\n  \"transportationInfo\": \"$inputRoot.transportationInfo\",\n  \"shortDescription\": \"$inputRoot.shortDescription\"\n}"
```
Comment (melon-ruet) [inline openapi/integration/aws/donations/post-donation.json:19]: This is not readable

**Review by melon-ruet — CHANGES_REQUESTED** [review]

File: openapi/integration/aws/donations/get-donations.json (hunk 2 above)
Comment (sakibstark11) [reply openapi/integration/aws/donations/get-donations.json:15]: are we using jsonencode for policies?

Comment (mahinuzzaman) [reply openapi/integration/aws/donations/get-donations.json:15]: we have to add string here. its not readable but is is the only way so far. we can create a vtl file but we have to create a script to handle the files

**Review by melon-ruet — DISMISSED** [review]

File: openapi/integration/aws/donations/get-donations.json (hunk 2 above)
Comment (mahinuzzaman) [reply openapi/integration/aws/donations/get-donations.json:15]: > are we using jsonencode for policies? 
> 

We are not. This time I have forgotten about this.

**Review by melon-ruet — APPROVED** [review]

**Pull request merged**